File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        
        s = s.replace("(", "0").replace(")", "1").replace("[", "2").replace("]", "3").replace("{", "4").replace("}", "5");
        
        int answer = 0;
        
        for (int i = 0; i < s.length(); i++) {
            s = s.substring(1) + s.charAt(0);
            answer += exec(s);
        }
        
        return answer;
    }
    
    public int exec(String s) {
        Deque<Integer> stack = new LinkedList<>();
        
        for (int i = 0; i < s.length(); i++) {     
            char c = s.charAt(i);
            if (c == '1' || c == '3' || c == '5') {
                int target = Integer.valueOf(c) - 1;
                if (stack.isEmpty() || stack.peekLast() != target){
                    return 0;
                }
                stack.pollLast();
            } else {
                stack.add(Integer.valueOf(c));
            }
        }
        
        if (stack.isEmpty()) {
            return 1;
        }
        return 0;
    }
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
# 회전을 통해 s를 올바른 문자열로 만든다
# s안에 올바른 문자열이 몇 개 있는지 확인한다

from collections import deque
def solution(s):
    
    for _ in range(len(s)):
        if not is_right(s):
            s = s[1:] + s[0]

    if not is_right(s):
        return 0
    
    print(s)
    left = 0
    right = 1
    answer = 0
    
    while right < len(s):
        if is_right(s[left:right+1]):
            answer += 1
            left = right + 1
            right = left + 1
            continue
        
        right += 1
    
    return answer
    
            
        
        

def is_right(s):
    q = deque()
    
    for c in s:
        if not q:
            if c == ")" or c == "}" or c == "]":
                return False
            q.append(c)
            continue
        
        if c == ")" and q[-1] == "(":
            q.pop()
        elif c == "}" and q[-1] == "{":
            q.pop()
        elif c == "]" and q[-1] == "[":
            q.pop()
        else:
            q.append(c)
    
    return not q
        
        
        
        
            
        
